pastis/_dispersion/__init__.py

```diff
@@ -0,0 +1,361 @@
+
+            t = lengths - i
+            # TODO check that this is correct...
+            mean_ = np.nansum(counts_ / bias_) / (t[t > 0].sum())
+            mean.append(mean_)
+
+            number_ = (t[t > 0]).sum() - len(counts_)
+
+            v = (np.nansum((counts_ / bias_ - mean_) ** 2) +
+                 np.nansum((mean_ * np.ones(number_)) ** 2))
+            v /= (t[t > 0].sum() - 1)
+
+            # Correct for the bias with all the biases of that genomic
+            # distance.
+            row, col = _get_indices_genomic_distances(lengths, i)
+            v /= np.nanmean(1. * (bias[row] * bias[col]))
+
+            var.append(v)
+            num_data_points.append(t[t > 0].sum())
+    return mean, var, num_data_points
+
+
+def estimate_dispersion(counts, lengths, bias=None,
+                        use_zero_counts=True, f=2./3, aggregation=stats.gmean):
+    """
+    Estimates the dispersion parameter assuming it is constant, using the
+    Geometric mean.
+
+    Parameters
+    ----------
+    counts : array-like
+
+    lengths : ndarray
+    """
+    # First, estimate unbiased mean and variances for all genomic distances of
+    # interest.
+    mean, var, _ = _compute_unbiased_mean_variance(
+        counts, lengths,
+        use_zero_counts=use_zero_counts, f=f,
+        bias=bias, return_num_data_points=True)
+
+    dispersions = mean ** 2 / (var - mean)
+    return stats.gmean(np.abs(dispersions))
+
+
+def _estimate_exponential_pol(mean, variance, degree=2, sample_weights=None,
+                              loss="linear"):
+    """
+    """
+    # Remove all points with mean == 0 or variance == 0
+    mask = ~((mean == 0) | (variance == 0))
+
+    # Stabilization
+    variance += 1e-8 * np.random.rand(variance.shape[0])
+    dispersion = mean[mask] ** 2 / (variance[mask] - mean[mask])
+    mean = mean[mask]
+    if sample_weights is not None:
+        sample_weights = sample_weights[mask]
+    else:
+        sample_weights = np.ones(mean.shape)
+
+    def eval_f(coef):
+        f = sum([coef[i] * np.log(mean) ** (i)
+                 for i in np.arange(degree+1)])
+        return sample_weights * (np.exp(f) - dispersion)
+
+    optimization_results = optimize.least_squares(
+        eval_f, np.zeros(degree+1), loss=loss)
+    coefficients = optimization_results["x"]
+    return coefficients
+
+
+class ExponentialDispersion(object):
+    """
+    """
+
+    def __init__(self, degree=2):
+        self.degree = degree
+
+    def fit(self, mean, variance, sample_weights=None):
+        self.coef_ = _estimate_exponential_pol(
+            mean, variance,
+            degree=self.degree,
+            sample_weights=sample_weights)
+        return self
+
+    def predict(self, mean, bias=None):
+        log_dispersion = np.zeros(mean.shape)
+        for i in np.arange(self.degree + 1):
+            log_dispersion += self.coef_[i] * np.log(mean) ** (i)
+
+        # Remove degenerarcies
```

Comment:
can have under/overflow in very very weird cases (extremely sparse and noisy simulation)